packages/aws-cdk/lib/commands/init/init.ts

```diff
@@ -76,8 +113,164 @@ export async function cliInit(options: CliInitOptions) {
+  }
+
+  return language;
+}
+
+/**
+ * Get valid CDK language directories from a template path
+ * @param templatePath - Path to the template directory
+ * @returns Promise resolving to array of supported language names
+ */
+async function getLanguageDirectories(templatePath: string): Promise<string[]> {
+  const cdkSupportedLanguages = ['typescript', 'javascript', 'python', 'java', 'csharp', 'fsharp', 'go'];
+  const languageExtensions: Record<string, string[]> = {
+    typescript: ['.ts'],
+    javascript: ['.js'],
+    python: ['.py'],
+    java: ['.java'],
+    csharp: ['.cs'],
+    fsharp: ['.fs'],
+    go: ['.go'],
+  };
+
+  try {
+    const entries = await fs.readdir(templatePath, { withFileTypes: true });
+
+    const languageValidationPromises = entries
+      .filter(directoryEntry => directoryEntry.isDirectory() && cdkSupportedLanguages.includes(directoryEntry.name))
+      .map(async (directoryEntry) => {
+        const languageDirectoryPath = path.join(templatePath, directoryEntry.name);
+        try {
+          const hasValidLanguageFiles = await hasLanguageFiles(languageDirectoryPath, languageExtensions[directoryEntry.name]);
+          return hasValidLanguageFiles ? directoryEntry.name : null;
+        } catch (error: any) {
+          throw new ToolkitError(`Cannot read language directory '${directoryEntry.name}': ${error.message}`);
+        }
+      });
+
+    // eslint-disable-next-line @cdklabs/promiseall-no-unbounded-parallelism
+    const validationResults = await Promise.all(languageValidationPromises);
+    return validationResults.filter((languageName): languageName is string => languageName !== null);
+  } catch (error: any) {
+    if (error instanceof ToolkitError) {
+      throw error;
+    }
+    throw new ToolkitError(`Cannot read template directory '${templatePath}': ${error.message}`);
+  }
+}
+
+/**
+ * Iteratively check if a directory contains files with the specified extensions
+ * @param directoryPath - Path to search for language files
+ * @param extensions - Array of file extensions to look for
+ * @returns Promise resolving to true if language files are found
+ */
+async function hasLanguageFiles(directoryPath: string, extensions: string[]): Promise<boolean> {
+  const dirsToCheck = [directoryPath];
+
+  while (dirsToCheck.length > 0) {
+    const currentDir = dirsToCheck.pop()!;
+
+    try {
+      const entries = await fs.readdir(currentDir, { withFileTypes: true });
+
+      for (const entry of entries) {
+        if (entry.isFile() && extensions.some(ext => entry.name.endsWith(ext))) {
+          return true;
+        } else if (entry.isDirectory()) {
+          dirsToCheck.push(path.join(currentDir, entry.name));
+        }
+      }
+    } catch (error: any) {
+      // Skip directories that can't be read (permissions, broken symlinks, etc.)
+      // but continue searching other directories
+      if (error.code === 'ENOENT' || error.code === 'EACCES' || error.code === 'ENOTDIR') {
```

Comment:
Why would we encounter ENOTDIR if we have a check for it on line 253?

Reply:
I also would not expect to see ENOENT or EACCES, should probably fail if there are not sufficient access rights.